# Data Analysis with Python

### Import required libraries

In [1]:
import pandas as pd
import numpy as np

### How to read an excel file?

In [2]:
df = pd.read_excel("test_cricket.xlsx", sheet_name='wickets')

display(df)

Player       Span  Mat  Inns  Balls   Runs  Wkts  \
0   M Muralitharan (ICC/SL)  1992-2010  133   230  44039  18180   800   
1            SK Warne (AUS)  1992-2007  145   273  40705  17995   708   
2          A Kumble (INDIA)  1990-2008  132   236  40850  18355   619   
3         JM Anderson (ENG)  2003-2021  162   301  34791  16457   617   
4          GD McGrath (AUS)  1993-2007  124   243  29248  12186   563   
..                      ...        ...  ...   ...    ...    ...   ...   
74        SCG MacGill (AUS)  1998-2008   44    85  11237   6038   208   
75    Saqlain Mushtaq (PAK)  1995-2004   49    86  14070   6206   208   
76         AME Roberts (WI)  1974-1983   47    90  11135   5174   202   
77            JA Snow (ENG)  1965-1976   49    93  12021   5387   202   
78         JR Thomson (AUS)  1972-1985   51    90  10535   5601   200   

                    BBI                  BBM    Ave  Econ    SR   5  10  
0   1951-09-01 00:00:00               16/220  22.72  2.47  55.0  67  22  
1   1971-08-01 00:00:00               12/128  25.41  2.65  57.4  37  10  
2   1974-10-01 00:00:00               14/149  29.65  2.69  65.9  35   8  
3   1942-07-01 00:00:00  1971-11-01 00:00:00  26.67  2.83  56.3  30   3  
4   2021-08-24 00:00:00  2021-10-27 00:00:00  21.64  2.49  51.9  29   3  
..                  ...                  ...    ...   ...   ...  ..  ..  
74                8/108               12/107  29.02  3.22  54.0  12   2  
75                8/164               10/155  29.83  2.64  67.6  13   3  
76  1954-07-01 00:00:00               12/121  25.61  2.78  55.1  11   2  
77  1940-07-01 00:00:00               10/142  26.66  2.68  59.5   8   1  
78  1946-06-01 00:00:00                9/105  28.00  3.18  52.6   8   0  

[79 rows x 14 columns]

### Displaying  the first 10 rows of the dataframe

In [3]:
df.head(10)

Player       Span  Mat  Inns  Balls   Runs  Wkts  \
0  M Muralitharan (ICC/SL)  1992-2010  133   230  44039  18180   800   
1           SK Warne (AUS)  1992-2007  145   273  40705  17995   708   
2         A Kumble (INDIA)  1990-2008  132   236  40850  18355   619   
3        JM Anderson (ENG)  2003-2021  162   301  34791  16457   617   
4         GD McGrath (AUS)  1993-2007  124   243  29248  12186   563   
5          SCJ Broad (ENG)  2007-2021  148   272  29713  14502   523   
6            CA Walsh (WI)  1984-2001  132   242  30019  12688   519   
7            DW Steyn (SA)  2004-2019   93   171  18608  10077   439   
8      N Kapil Dev (INDIA)  1978-1994  131   227  27740  12867   434   
9        HMRKB Herath (SL)  1999-2018   93   170  25993  12157   433   

                   BBI                  BBM    Ave  Econ    SR   5  10  
0  1951-09-01 00:00:00               16/220  22.72  2.47  55.0  67  22  
1  1971-08-01 00:00:00               12/128  25.41  2.65  57.4  37  10  
2  1974-10-01 00:00:00               14/149  29.65  2.69  65.9  35   8  
3  1942-07-01 00:00:00  1971-11-01 00:00:00  26.67  2.83  56.3  30   3  
4  2021-08-24 00:00:00  2021-10-27 00:00:00  21.64  2.49  51.9  29   3  
5  2021-08-15 00:00:00               11/121  27.72  2.92  56.8  18   3  
6  1937-07-01 00:00:00                13/55  24.44  2.53  57.8  22   3  
7  1951-07-01 00:00:00  1960-11-01 00:00:00  22.95  3.24  42.3  26   5  
8  1983-09-01 00:00:00               11/146  29.64  2.78  63.9  23   2  
9                9/127               14/184  28.07  2.80  60.0  34   9

### Explaining the meaning of each column.
<br> **Player**- Name of the player
<br>  **Span** - Career time period of the player
<br>  **Mat** - Number of test matches played.
<br>  **Inns** - Number of innings played.
<br>  **Balls** - Number of balls played.
<br> **BBI** - Number of best bowling innings
<br>  **BBM** - Best bowling in match
<br>  **Ave** Bowling average
<br>  **Econ** - Economy rate
<br> **SR** - Strike rate
<br>  **5** - Number of times he took 5-9 wickets.
<br>  **10** - Number of times he took 10 wickets.



### Finding the number of rows and columns in the dataframe

In [4]:
# Rows
print("number of rows = ", df.shape[0])

# Columns
print("number of columns = ", df.shape[1])

number of rows =  79
number of columns =  14


 ### Finding the data statistics and checking for the data types

In [5]:
# checking data statistics
display(df.describe())

# checking for the data types
df.dtypes

Mat        Inns         Balls          Runs        Wkts  \
count   79.000000   79.000000     79.000000     79.000000   79.000000   
mean    80.101266  144.797468  18630.303797   8595.506329  317.101266   
std     28.537692   51.042310   7190.036515   3080.256645  121.731587   
min     37.000000   67.000000   8785.000000   4846.000000  200.000000   
25%     60.500000  110.000000  13580.000000   6456.500000  229.000000   
50%     71.000000  129.000000  16498.000000   7742.000000  266.000000   
75%     93.000000  169.000000  21742.500000   9756.000000  374.500000   
max    166.000000  301.000000  44039.000000  18355.000000  800.000000   

             Ave       Econ         SR          5         10  
count  79.000000  79.000000  79.000000  79.000000  79.000000  
mean   27.466456   2.806582  59.187342  16.354430   2.797468  
std     3.657561   0.351666   9.349337   9.642372   3.235935  
min    20.940000   1.980000  41.200000   3.000000   0.000000  
25%    24.425000   2.600000  53.300000   9.500000   1.000000  
50%    28.000000   2.820000  57.400000  14.000000   2.000000  
75%    29.870000   3.080000  63.950000  20.500000   3.500000  
max    34.790000   3.460000  91.900000  67.000000  22.000000

Player     object
Span       object
Mat         int64
Inns        int64
Balls       int64
Runs        int64
Wkts        int64
BBI        object
BBM        object
Ave       float64
Econ      float64
SR        float64
5           int64
10          int64
dtype: object

**Overview:**
1. The average number of  matches played by the top bowlers is 80, with minimum of 37 matches and maximum of 166.
2. The average number of  innings played by the top bowlers is 144, with minimum of 67 innings and maximum of 301.
3. The average number of  balls played by the top bowlers is 18630, with minimum of 8785 balls and maximum of 44039.
4. The average number of runs made by the bowlers is 8595. While the highest runs is 18355, only 25% scored more than 9756 runs. 
5. The average number of wickets taken by the bowlers is 317, with 75% of bowlers scoring less than 375.

### Checking for missing values in the dataset

In [6]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Player  79 non-null     object 
 1   Span    79 non-null     object 
 2   Mat     79 non-null     int64  
 3   Inns    79 non-null     int64  
 4   Balls   79 non-null     int64  
 5   Runs    79 non-null     int64  
 6   Wkts    79 non-null     int64  
 7   BBI     79 non-null     object 
 8   BBM     79 non-null     object 
 9   Ave     79 non-null     float64
 10  Econ    79 non-null     float64
 11  SR      79 non-null     float64
 12  5       79 non-null     int64  
 13  10      79 non-null     int64  
dtypes: float64(3), int64(7), object(4)
memory usage: 8.8+ KB
None


**Overview:** There is no missing value in the dataset.

### Renaming columns

In [7]:
df = df.rename(columns={'Mat':'Matches', 
                        'Inns':'Innings',
                        'Wkts': 'Wickets',
                        'Ave': 'Average',
                        'Econ': 'Economy'})

display(df.head())

Player       Span  Matches  Innings  Balls   Runs  \
0  M Muralitharan (ICC/SL)  1992-2010      133      230  44039  18180   
1           SK Warne (AUS)  1992-2007      145      273  40705  17995   
2         A Kumble (INDIA)  1990-2008      132      236  40850  18355   
3        JM Anderson (ENG)  2003-2021      162      301  34791  16457   
4         GD McGrath (AUS)  1993-2007      124      243  29248  12186   

   Wickets                  BBI                  BBM  Average  Economy    SR  \
0      800  1951-09-01 00:00:00               16/220    22.72     2.47  55.0   
1      708  1971-08-01 00:00:00               12/128    25.41     2.65  57.4   
2      619  1974-10-01 00:00:00               14/149    29.65     2.69  65.9   
3      617  1942-07-01 00:00:00  1971-11-01 00:00:00    26.67     2.83  56.3   
4      563  2021-08-24 00:00:00  2021-10-27 00:00:00    21.64     2.49  51.9   

    5  10  
0  67  22  
1  37  10  
2  35   8  
3  30   3  
4  29   3

### Removing columns

In [8]:
df.drop('BBI', axis=1, inplace=True)

display(df.head())

Player       Span  Matches  Innings  Balls   Runs  \
0  M Muralitharan (ICC/SL)  1992-2010      133      230  44039  18180   
1           SK Warne (AUS)  1992-2007      145      273  40705  17995   
2         A Kumble (INDIA)  1990-2008      132      236  40850  18355   
3        JM Anderson (ENG)  2003-2021      162      301  34791  16457   
4         GD McGrath (AUS)  1993-2007      124      243  29248  12186   

   Wickets                  BBM  Average  Economy    SR   5  10  
0      800               16/220    22.72     2.47  55.0  67  22  
1      708               12/128    25.41     2.65  57.4  37  10  
2      619               14/149    29.65     2.69  65.9  35   8  
3      617  1971-11-01 00:00:00    26.67     2.83  56.3  30   3  
4      563  2021-10-27 00:00:00    21.64     2.49  51.9  29   3

### Removing BBM Column


In [9]:
df.drop('BBM', axis=1, inplace=True)

display(df.head())


Player       Span  Matches  Innings  Balls   Runs  \
0  M Muralitharan (ICC/SL)  1992-2010      133      230  44039  18180   
1           SK Warne (AUS)  1992-2007      145      273  40705  17995   
2         A Kumble (INDIA)  1990-2008      132      236  40850  18355   
3        JM Anderson (ENG)  2003-2021      162      301  34791  16457   
4         GD McGrath (AUS)  1993-2007      124      243  29248  12186   

   Wickets  Average  Economy    SR   5  10  
0      800    22.72     2.47  55.0  67  22  
1      708    25.41     2.65  57.4  37  10  
2      619    29.65     2.69  65.9  35   8  
3      617    26.67     2.83  56.3  30   3  
4      563    21.64     2.49  51.9  29   3

### Finding how many players played for ICC

In [10]:
# For this we will need to seperate the country column from the player
df = df.rename(columns={'SR': 'Strike_Rate'})
# splitting the 'Player' column to get the information about 'Country'
df_player = df['Player'].str.split("(", expand=True)

# concatinating 'Country' with the main dataframe
df = pd.concat([df, df_player], axis=1)

# dropping the 'Player' columns
df = df.drop('Player', axis=1)

# renaming the column names
df = df.rename(columns={0: 'Player',
                        1: 'Country'})

# remove the ")" from the 'Country' column
df['Country'] = df['Country'].str.replace(")", "")

# rearrange the columns
new_col_sequence = ['Player', 'Span', 'Matches', 'Innings', 'Runs', 'Wickets',
                    'Average', 'Economy', 'Strike_Rate', 5, 'Country']
df = df[new_col_sequence]

display(df.head())

<ipython-input-10-194cf972045c>:17: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  df['Country'] = df['Country'].str.replace(")", "")


Player       Span  Matches  Innings   Runs  Wickets  Average  \
0  M Muralitharan   1992-2010      133      230  18180      800    22.72   
1        SK Warne   1992-2007      145      273  17995      708    25.41   
2        A Kumble   1990-2008      132      236  18355      619    29.65   
3     JM Anderson   2003-2021      162      301  16457      617    26.67   
4      GD McGrath   1993-2007      124      243  12186      563    21.64   

   Economy  Strike_Rate   5 Country  
0     2.47         55.0  67  ICC/SL  
1     2.65         57.4  37     AUS  
2     2.69         65.9  35   INDIA  
3     2.83         56.3  30     ENG  
4     2.49         51.9  29     AUS

In [11]:
# defining the function with condition
def icc_check(x):
    if "ICC" in x:
        return "Yes"
    else:
        return "No"


In [12]:
df['played_for_ICC'] = df['Country'].apply(icc_check)


In [13]:
df['played_for_ICC'].value_counts()

No     74
Yes     5
Name: played_for_ICC, dtype: int64

### Finding number of different countries are present in this dataset

In [14]:
df['Country'].value_counts()

AUS        18
ENG        13
INDIA      10
WI          9
PAK         7
SA          7
NZ          6
SL          2
ENG/ICC     2
BDESH       1
ICC/SA      1
ICC/SL      1
ICC/NZ      1
ZIM         1
Name: Country, dtype: int64

### Finding the player who had played for the longest period of time

In [15]:
# splitting the 'Span' column based on the "-"
df_span = df['Span'].str.split("-", expand=True)

# conccatinating the new dataframe with the main dataframe
df = pd.concat([df, df_span], axis=1)

# renaming the newly created column names
df = df.rename(columns={0: "start_year",
                        1: "end_year"})

# removing the "Span" column
df = df.drop("Span", axis=1)

display(df.head())

Player  Matches  Innings   Runs  Wickets  Average  Economy  \
0  M Muralitharan       133      230  18180      800    22.72     2.47   
1        SK Warne       145      273  17995      708    25.41     2.65   
2        A Kumble       132      236  18355      619    29.65     2.69   
3     JM Anderson       162      301  16457      617    26.67     2.83   
4      GD McGrath       124      243  12186      563    21.64     2.49   

   Strike_Rate   5 Country played_for_ICC start_year end_year  
0         55.0  67  ICC/SL            Yes       1992     2010  
1         57.4  37     AUS             No       1992     2007  
2         65.9  35   INDIA             No       1990     2008  
3         56.3  30     ENG             No       2003     2021  
4         51.9  29     AUS             No       1993     2007

In [16]:
#checking the datatype
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Player          79 non-null     object 
 1   Matches         79 non-null     int64  
 2   Innings         79 non-null     int64  
 3   Runs            79 non-null     int64  
 4   Wickets         79 non-null     int64  
 5   Average         79 non-null     float64
 6   Economy         79 non-null     float64
 7   Strike_Rate     79 non-null     float64
 8   5               79 non-null     int64  
 9   Country         79 non-null     object 
 10  played_for_ICC  79 non-null     object 
 11  start_year      79 non-null     object 
 12  end_year        79 non-null     object 
dtypes: float64(3), int64(5), object(5)
memory usage: 8.1+ KB


In [17]:
#converting data type
df['start_year'] = df['start_year'].astype('int') 
df['end_year'] = df['end_year'].astype('int')

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Player          79 non-null     object 
 1   Matches         79 non-null     int64  
 2   Innings         79 non-null     int64  
 3   Runs            79 non-null     int64  
 4   Wickets         79 non-null     int64  
 5   Average         79 non-null     float64
 6   Economy         79 non-null     float64
 7   Strike_Rate     79 non-null     float64
 8   5               79 non-null     int64  
 9   Country         79 non-null     object 
 10  played_for_ICC  79 non-null     object 
 11  start_year      79 non-null     int32  
 12  end_year        79 non-null     int32  
dtypes: float64(3), int32(2), int64(5), object(3)
memory usage: 7.5+ KB
None


In [18]:
# Making a column for years the player played
df['years_played'] = df['end_year'] - df['start_year']

df = df.drop(['start_year', "end_year"], axis=1)

display(df.head(10))

Player  Matches  Innings   Runs  Wickets  Average  Economy  \
0  M Muralitharan       133      230  18180      800    22.72     2.47   
1        SK Warne       145      273  17995      708    25.41     2.65   
2        A Kumble       132      236  18355      619    29.65     2.69   
3     JM Anderson       162      301  16457      617    26.67     2.83   
4      GD McGrath       124      243  12186      563    21.64     2.49   
5       SCJ Broad       148      272  14502      523    27.72     2.92   
6        CA Walsh       132      242  12688      519    24.44     2.53   
7        DW Steyn        93      171  10077      439    22.95     3.24   
8     N Kapil Dev       131      227  12867      434    29.64     2.78   
9    HMRKB Herath        93      170  12157      433    28.07     2.80   

   Strike_Rate   5 Country played_for_ICC  years_played  
0         55.0  67  ICC/SL            Yes            18  
1         57.4  37     AUS             No            15  
2         65.9  35   INDIA             No            18  
3         56.3  30     ENG             No            18  
4         51.9  29     AUS             No            14  
5         56.8  18     ENG             No            14  
6         57.8  22      WI             No            17  
7         42.3  26      SA             No            15  
8         63.9  23   INDIA             No            16  
9         60.0  34      SL             No            19

### Top 5 batsmen: Highest number of years played

In [19]:
df.sort_values(by="years_played", ascending = False).head(5)

Player  Matches  Innings   Runs  Wickets  Average  Economy  \
21      Imran Khan        88      142   8258      362    22.81     2.54   
55       GS Sobers        93      159   7999      235    34.03     2.22   
9     HMRKB Herath        93      170  12157      433    28.07     2.80   
0   M Muralitharan       133      230  18180      800    22.72     2.47   
2         A Kumble       132      236  18355      619    29.65     2.69   

    Strike_Rate   5 Country played_for_ICC  years_played  
21         53.7  23     PAK             No            21  
55         91.9   6      WI             No            20  
9          60.0  34      SL             No            19  
0          55.0  67  ICC/SL            Yes            18  
2          65.9  35   INDIA             No            18

### Top 5 batsmen: Lowest number of years played

In [20]:
df.sort_values(by="years_played", ascending = True).head(5)

Player  Matches  Innings  Runs  Wickets  Average  Economy  \
44      GP Swann        60      109  7642      255    29.96     2.98   
71      K Rabada        47       86  4846      213    22.75     3.30   
54    Yasir Shah        45       83  7247      235    30.83     3.19   
61   SJ Harmison        63      115  7192      226    31.82     3.22   
72  JR Hazlewood        55      103  5438      212    25.65     2.74   

    Strike_Rate   5  Country played_for_ICC  years_played  
44         60.1  17      ENG             No             5  
71         41.2  10       SA             No             6  
54         57.8  16      PAK             No             7  
61         59.1   8  ENG/ICC            Yes             7  
72         56.0   9      AUS             No             7

### Number of Australian Bowlers present in this dataset

In [21]:
# defining the function with condition
def aus__player_check(x):
    if "AUS" in x:
        return "Australian"
    else:
        return "Not Australian"


In [22]:
df['australian_bowler'] = df['Country'].apply(aus__player_check)

In [23]:
df['australian_bowler'].value_counts()

Not Australian    61
Australian        18
Name: australian_bowler, dtype: int64

### Finding for Bangladeshi player in the dataset

In [24]:
# defining the function with condition
def bengali__player_check(x):
    if "BAN" in x:
        return "Bengali Players Found"
    else:
        return "Not found"

In [25]:
df['bengali_player'] = df['Country'].apply(bengali__player_check)

In [26]:
df['bengali_player'].value_counts()

Not found    79
Name: bengali_player, dtype: int64

### Finding player with lowest economy rate 

In [27]:
df.sort_values(by="Economy", ascending = True).head(1)

Player  Matches  Innings  Runs  Wickets  Average  Economy  Strike_Rate  \
32  LR Gibbs        79      148  8989      309    29.09     1.98         87.7   

     5 Country played_for_ICC  years_played australian_bowler bengali_player  
32  18      WI             No            18    Not Australian      Not found

### Finding player with lowest strike rate 

In [28]:
df.sort_values(by="Strike_Rate", ascending = True).head(1)

Player  Matches  Innings  Runs  Wickets  Average  Economy  Strike_Rate  \
71  K Rabada        47       86  4846      213    22.75      3.3         41.2   

     5 Country played_for_ICC  years_played australian_bowler bengali_player  
71  10      SA             No             6    Not Australian      Not found

### Finding player with lowest bowling average

In [29]:
df.sort_values(by="Average", ascending = True).head(1)

Player  Matches  Innings  Runs  Wickets  Average  Economy  \
19  MD Marshall        81      151  7876      376    20.94     2.68   

    Strike_Rate   5 Country played_for_ICC  years_played australian_bowler  \
19         46.7  22      WI             No            13    Not Australian   

   bengali_player  
19      Not found